docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+
+- many kubectl commands depend on internal kubernetes/kubernetes libraries
+- most commands should not have these dependencies, but removing them requires large rewrites of the commands
+- *some* commands do actually need these dependencies (e.g. convert)
+- continuing to develop in the kubernetes/kubernetes repo results in more *bad* dependencies being added
+  even as we try to remove old ones
+- many commands depend on test infrastructure bespoke to kubernetes/kubernetes, which would need to be moved as well
+
+Additionally, many kubectl commands are laden with technical debt, using anti-patterns for working with APIs
+that do not work with version skew or support extensibility.  We have since grown out of using these patterns,
+but they are still pervasive.  Frequently, is it much faster and effective to rewrite large pieces
+instead of trying to refactor them into different designs.
+
+Some examples of anti-patterns:
+
+- Relying on specific compiled in API types for generic operations (scale, rollback)
+- Using internal type definitions (printers, various other commands)
+- Round tripping objects read from the server through go structures (some exceptions where this maybe ok)
+
+## Goals for ktl
+
+Thing we need.
+
+Keep the advantages of:
+
+- easy distribution
+- reasonably sized (~100MB)
+- easy installation
+- use common cli / client infrastructure
+- discoverable commands
+
+And also:
+
+- allow the cli to be released independently from the cluster
+- allow sub command groups within the cli to be released independently from one another
+- allow a decentralized ecosystem of tools to leverage centralized maintained cli / client infrastructure
+- facilitate end-to-end ownership of the cli, and in some cases sub command groups with the cli
+- facilitate decentralized development of extensions for the cli
+
+## Anti-goals
+
+Things we want to avoid.
+
+- block on moving existing kubectl commands out of the kubernetes/kubernetes repo
+- rewrite kubectl from the ground up in a new repo
+
+## Non-goals
+
+Even if these are good ideas, don't let them distract us from meeting our goals will simpler solutions.
+
+- build solution for discovering installable plugins and installing them
+  - rely on existing package management solutions for this until we need something more
+- invent new build and distribution infrastructure
+- fix issues with the existing kubectl commands
+- dog fooding the plugin mechanism for core commands
+
+## Reference-level explanation
+
+- build a new cli binary `ktl` (kubectl2) under the `kubernetes/ktl` repo that dispatches to commands developed in other repos
+- keep old cli commands in `kubernetes/kubernetes/cmd/kubectl` and vendor them into `kubectl/ktl`
+- build new cli commands in `kubernetes/kubectl` and vendor them into `kurnetes/ktl`
+- build common cli Kubernetes client infrastructure and libraries that can be used to develop a decentralized cli ecosystem
```

Comment:
You may want to clarify where the CLI infra code lives. I assume it is `kubernetes/kubectl`.